packages/byods/src/BYODS.ts

```diff
@@ -1,47 +1,389 @@
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${token}`,
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          targetOrgId: orgId,
+          clientId: this.config.clientId,
+          clientSecret: this.config.clientSecret,
+        }),
+      });
 
-    // Use this.token.access_token to make authenticated requests
-    const response = await fetch(endpoint, {
+      if (!response.ok) {
+        throw new Error(`Failed to retrieve token: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+    } catch (error) {
+      console.error('Error retrieving token after authorization:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Get the service app authorization for a given organization ID.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<ServiceAppAuthorization>}
+   */
+  public async getServiceAppAuthorization(orgId: string): Promise<ServiceAppAuthorization> {
+    if (!this.serviceAppAuthorizations[orgId]) {
+      return Promise.reject(new Error('Service app authorization not found'));
+    }
+
+    return Promise.resolve(this.serviceAppAuthorizations[orgId]);
+  }
+
+  /**
+   * Refresh the access token using the refresh token.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<void>}
+   */
+  public async refreshServiceAppAccessToken(orgId: string): Promise<void> {
+    const serviceAppAuthorization = await this.getServiceAppAuthorization(orgId);
+    const refreshToken = serviceAppAuthorization?.token.refreshToken;
+    if (!refreshToken) {
+      throw new Error('Refresh token is undefined');
+    }
+    try {
+      const response = await fetch(`${this.baseUrl}/access_token`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          grant_type: 'refresh_token',
+          client_id: this.config.clientId,
+          client_secret: this.config.clientSecret,
+          refresh_token: refreshToken,
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to refresh access token: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+    } catch (error) {
+      console.error('Error refreshing access token:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Make an authenticated request to the specified endpoint.
+   * @param {string} orgId - The organization ID.
+   * @param {string} endpoint - The endpoint to make the request to.
+   * @param {RequestInit} [options={}] - The request options.
+   * @returns {Promise<ApiResponse<T>>}
+   */
+  public async request<T>(
```

Comment:
Do we not know the return type here or are we just doing this to keep the type generic?

Reply:
This is meant to be a generic templated request. This can take any shape of data. Concrete type will be provided in the methods that call this